### DISCUSSION TITLE
addFace() error and strange list produced with some strange alien numbers

### DISCUSSION THREAD

**Discussion by blower05** [InteractiveComputerGraphics/PositionBasedDynamics#134] [Q&A] [answered]
Hi Bender, 

I tried to do some simulations of gear collision by inputting their geometry by initializing an instance of pbd.IndexedFaceMesh and  use addFace(). However, I got the following error.

TypeError: addFace(): incompatible function arguments. The following argument types are supported:
    1. (self: pypbd.IndexedFaceMesh, arg0: int) -> None
    2. (self: pypbd.IndexedFaceMesh, arg0: int) -> None

Invoked with: <pypbd.IndexedFaceMesh object at 0x000001E97C958F30>, 1, 1, 2
Error: Python script failed, check the message in the system console

Furthermore,  I, otherwise, tried inputting a range of list like the following:
```
for i in [1, 1, 2, 2, 3,3, 3,3, 2,2, 4, 4]:
        mes.verts.ensure_lookup_table()
        mesh.addFace(i ) 

print( mesh.getFaces().tolist() )
```

However, it returned the following strange big number for every alternate index positions 1 and  2:
 [1, 489, 270402240, 1, 489, 270402240, 2, 489, 270402240, 2, 489, 270402240, 3, 489, 270402240, 3, 489, 270402240, 3, 489]
I don't how to input the data to replace the strange 489 and 270402240 

Would you please tell me how to input the arguments and format the data beforehand?

Thanks

**Comment by janbender** [comment]
The function addFace requires 3 indices per triangle. However, you call it with just one int. So probably the remaining two indices are just random numbers. Can you try to call the function with something like this: 

addFace([1,2,3])

**Reply by blower05** [reply]
Thanks for the prompt reply!
I did try to input like what you suggested. However, it still get the same error below.
Also, I noted the arg0 is an integer rather than a list or other kind of types. So, I bet it should be some keywords arguments like I mentioned in the last post. I tried to dig out from the UtilitiesModule.cpp that it seems to give hints about the keywords argument should look like posIndices, texIndices and normalIndices. However, I still couldn't make it. Is there any other kw should be used that is hidden because no documentation mentioned about this. In addtions, it stated arg0: int it is probably some face indices but not vertices points because the are floating points.

TypeError: addFace(): incompatible function arguments. The following argument types are supported:
    1. (self: pypbd.IndexedFaceMesh, arg0: int) -> None
    2. (self: pypbd.IndexedFaceMesh, arg0: int) -> None

Invoked with: <pypbd.IndexedFaceMesh object at 0x0000026B4DCDDF70>, [1, 2, 3]
Error: Python script failed, check the message in the system console



Thanks

**Comment by janbender** [comment]
It seems to be a pointer. Probably I have to adapt the Python interface. I will take a look next week.

**Comment by blower05** [comment]
I tried hard and worked till now for many different possible combinations but in vain. Yes, thanks for your suggestion that it maybe a pointer for the integers. it looks python does not deal with the pointer, am I right?

**Comment by janbender** [comment, marked as answer]
I just fixed the Python interface of the function. Now you should be able to add a face by calling:

addFace([1,2,3])

This will add a triangle with the vertex indices 1,2 and 3.

**Reply by blower05** [reply]
Wow, Bender! That's very kind of you great!
Since I am now working on python 3.9, Instead of download again for the python 3.10, can I just download the mentioned UtilitiesModule.cpp file to overwrite the previous one to make it work?

Thanks

**Comment by janbender** [comment]
This should work. And of course then you have to recompile it. 
Alternatively just install it by: 

pip install pypbd

**Reply by blower05** [reply]
Yes. Thanks!
I tried to replaced the .cpp file and it didn't work but I found lately you have also kindly updated for the python 3.9 version and it is good for me and I am happy to continue to work it out now.
but I am now try to figuring out how to extract the tex and normal indices.

Before I can figure it out, I just try to put 0s for the tex and normal positions to test it out:
for i, f in enumerate(faces):
    mesh.addFace([f.index, 0, 0])
it returned the list:
 [0, 0, 0, 1, 0, 0, 2, 0, 0, 3, 0, 0, 4, 0, 0, 5, 0, 0]

Thanks
